src/psyclone/tests/psyir/frontend/fparser2_select_type_test.py

```diff
@@ -0,0 +1,415 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author R. W. Ford, STFC Daresbury Lab
+
+'''Module containing pytest tests for the handling of select type
+construction for the Fparser->PSyIR frontend.
+
+'''
+import pytest
+
+from fparser.common.readfortran import FortranStringReader
+from fparser.two import Fortran2003
+from fparser.two.parser import ParserFactory
+
+from psyclone.errors import InternalError
+from psyclone.psyir.frontend.fparser2 import Fparser2Reader
+from psyclone.psyir.nodes import CodeBlock, IfBlock
+from psyclone.tests.utilities import Compile
+
+
+def test_invalid_node():
+    '''Check that the expected exception is raised if the supplied node is
+    not an fparser2 Select_Type_Construct.
+
+    '''
+    reader = Fparser2Reader()
+    with pytest.raises(InternalError) as info:
+        reader._type_construct_handler(None, None)
+    assert ("Failed to find opening select type statement in: None"
+            in str(info.value))
+
+
+def test_no_end_select():
+    '''Check that the expected exception is raised if the supplied fparser
+    tree root node does not have End_Select_Type_Stmt as its last child.
+
+    '''
+    code = (
+        "  SELECT TYPE (type_selector)\n"
+        "    TYPE IS (INTEGER)\n"
+        "      branch1 = 1\n"
+        "      branch2 = 0\n"
+        "    TYPE IS (REAL)\n"
+        "      branch2 = 1\n"
+        "  END SELECT\n")
+    string_reader = FortranStringReader(code)
+    ParserFactory().create(std="f2008")
+    ast = Fortran2003.Select_Type_Construct(string_reader)
+    # Remove the end_select instance to force the exception
+    del ast.children[-1]
+    reader = Fparser2Reader()
+    with pytest.raises(InternalError) as info:
+        reader._type_construct_handler(ast, None)
+    assert("Failed to find closing select type statement in: "
```

Comment:
Also likely to cause pain and suffering in the future. (Well, OK, it will break.)

Reply:
Removed test.